pkg/frontend/admin_openshiftcluster_etcdcertificaterenew.go

```diff
@@ -0,0 +1,431 @@
+	}
+	return nil
+}
+
+func (e *etcdrenew) revertChanges(ctx context.Context) error {
+	s := []steps.Step{
+		steps.Action(e.fetchEtcdCurrentRevision),
+		steps.Action(e.recoverEtcdSecrets),
+		steps.Condition(e.isEtcdRevised, 30*time.Minute, true),
+	}
+	_, err := steps.Run(ctx, e.log, 10*time.Second, s, nil)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (f *frontend) _postAdminOpenShiftClusterEtcdCertificateRenew(ctx context.Context, r *http.Request, log *logrus.Entry) error {
+	resourceName := chi.URLParam(r, "resourceName")
+	resourceType := chi.URLParam(r, "resourceType")
+	resourceGroupName := chi.URLParam(r, "resourceGroupName")
+
+	resourceID := strings.TrimPrefix(r.URL.Path, "/admin")
+
+	doc, err := f.dbOpenShiftClusters.Get(ctx, resourceID)
+	switch {
+	case cosmosdb.IsErrorStatusCode(err, http.StatusNotFound):
+		return api.NewCloudError(http.StatusNotFound, api.CloudErrorCodeResourceNotFound, "", "The Resource '%s/%s' under resource group '%s' was not found.", resourceType, resourceName, resourceGroupName)
+	case err != nil:
+		return err
+	}
+
+	k, err := f.kubeActionsFactory(log, f.env, doc.OpenShiftCluster)
+	if err != nil {
+		return err
+	}
+	e := &etcdrenew{
+		log:           log,
+		k:             k,
+		secretNames:   nil,
+		mode:          "",
+		backupSecrets: make(map[string][]byte),
+		lastRevision:  0,
+	}
+
+	if err = e.validateClusterVersion(ctx); err != nil {
+		return err
+	}
+
+	// Fetch secretNames using nodeNames
+	masterNodeNames, err := fetchNodeNames(ctx, k, log)
+	if err != nil {
+		return err
+	}
+	e.log.Infof("Nodes found: %s", masterNodeNames)
+	if len(masterNodeNames) != 3 {
+		return api.NewCloudError(http.StatusForbidden, api.CloudErrorCodeForbidden, "", "The cluster doesn't have 3 master nodes, quitting.")
+	}
+	for _, nodeName := range masterNodeNames {
+		for _, prefix := range []string{"etcd-peer-", "etcd-serving-", "etcd-serving-metrics-"} {
+			e.secretNames = append(e.secretNames, prefix+nodeName)
+		}
+	}
+
+	// validate etcd and certificates, backup and delete secrets
+	if err = e.validateEtcdAndBackupDeleteSecretOnFlagSet(ctx, true); err != nil {
+		return err
+	}
+
+	// Once secrets are deleted, the operator recreates the secrets and a new etcd revision is applied
+	// On failure, proceed for recovery by applying the backupsecrets on the cluster again
+	// On success, verify the etcd state, certificates
+	err = e.isEtcDRootCertRenewed(ctx)
+	if err != nil {
+		e.log.Infoln("Attempting to recover from backup, and wait for new revision to be applied after recovery")
+		return e.revertChanges(ctx)
+	}
+
+	e.log.Infoln("Etcd certificates are renewed and new revision is applied, verifying.")
+	e.mode = "renewed"
+	return e.validateEtcdAndBackupDeleteSecretOnFlagSet(ctx, false)
+}
+
+func (e *etcdrenew) validateClusterVersion(ctx context.Context) error {
+	e.log.Infoln("validating cluster version now")
+	rawCV, err := e.k.KubeGet(ctx, "ClusterVersion.config.openshift.io", "", "version")
```

Comment:
nit, but the extra `name` in this call to KubeGet is optional:
```
oc get clusterversion.config.openshift.io
NAME      VERSION                              AVAILABLE   PROGRESSING   SINCE   STATUS
version   4.14.0-0.nightly-2023-08-08-222204   True        False         35d     Cluster version is 4.14.0-0.nightly-2023-08-08-222204

oc get clusterversion.config.openshift.io version
NAME      VERSION                              AVAILABLE   PROGRESSING   SINCE   STATUS
version   4.14.0-0.nightly-2023-08-08-222204   True        False         35d     Cluster version is 4.14.0-0.nightly-2023-08-08-222204
```

Reply:
I have added "version" to make it more readable :)